appmgr/src/disk/quirks.rs

```diff
@@ -0,0 +1,156 @@
+    }
+}
+impl Default for Quirks {
+    fn default() -> Self {
+        Quirks(vec![(VendorId(0x152d), ProductId(0x0562))]) // SSK enclosure shipped by Start9 Labs
+    }
+}
+impl std::fmt::Display for Quirks {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut comma = false;
+        for (vendor, product) in &self.0 {
+            if comma {
+                write!(f, ",")?;
+                comma = true;
+            }
+            write!(f, "{}:{}:u", vendor, product)?;
+        }
+        Ok(())
+    }
+}
+impl std::str::FromStr for Quirks {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s = s.trim();
+        let mut quirks = Vec::new();
+        for item in s.split(",") {
+            if let [vendor, product, "u"] = item.splitn(3, ":").collect::<Vec<_>>().as_slice() {
+                quirks.push((vendor.parse()?, product.parse()?));
+            } else {
+                return Err(Error::new(
+                    eyre!("Invalid quirk: `{}`", item),
+                    crate::ErrorKind::DiskManagement,
+                ));
+            }
+        }
+        Ok(Quirks(quirks))
+    }
+}
+
+#[instrument]
+pub async fn update_quirks(quirks: &mut Quirks) -> Result<(), Error> {
+    let mut usb_devices = tokio::fs::read_dir("/sys/bus/usb/devices/").await?;
+    while let Some(usb_device) = usb_devices.next_entry().await? {
+        if tokio::fs::metadata(usb_device.path().join("idVendor"))
+            .await
+            .is_err()
+        {
+            continue;
+        }
+        let vendor = tokio::fs::read_to_string(usb_device.path().join("idVendor"))
+            .await?
+            .parse()?;
+        let product = tokio::fs::read_to_string(usb_device.path().join("idProduct"))
+            .await?
+            .parse()?;
+        if WHITELIST.contains(&(vendor, product)) || quirks.contains(vendor, product) {
+            continue;
+        }
+        quirks.add(vendor, product);
+        Command::new("modprobe")
+            .arg("usb_storage")
+            .arg(format!("quirks={}", quirks))
+            .invoke(crate::ErrorKind::DiskManagement)
+            .await?;
+
+        // reconnect usb device
+        tokio::fs::write(usb_device.path().join("authorized"), "0").await?;
+        tokio::fs::write(usb_device.path().join("authorized"), "1").await?;
+    }
+    Ok(())
+}
+
+#[instrument]
+pub async fn fetch_quirks() -> Result<Quirks, Error> {
+    if tokio::fs::metadata(QUIRK_PATH).await.is_ok() {
+        Ok(tokio::fs::read_to_string(QUIRK_PATH).await?.parse()?)
+    } else {
+        Ok(Quirks::default())
+    }
+}
+
+#[instrument]
+pub async fn save_quirks(quirks: &Quirks) -> Result<(), Error> {
+    let mut quirk_file = AtomicFile::new(QUIRK_PATH).await?;
+    quirk_file.write_all(quirks.to_string().as_bytes()).await?;
+    quirk_file.save().await?;
+    let guard = TmpMountGuard::mount("/dev/mmcblk0p1", None).await?;
```

Comment:
This seems fragile. Means that none of this code will work for systems that don't use mmc for the installation. This is a hidden hardware dependency.

Reply:
I am planning a refactor to the mounts module as part of cloud backups where this will be abstracted. This will work for now.